RIOT-OS/RIOT
# nanocoap: cache for CoAP responses

### Contribution description
This PR adds a cache for CoAP response messages to `nanocoap`. [RFC7252, Section-5.6](https://tools.ietf.org/html/rfc7252#section-5.6) elaborates on the internals of a cache implementation.

The basic idea is to reduce response time and enhance bandwidth usage by storing CoAP responses at intermediary proxy nodes. The cache key is calculated using the information in [RFC7252, Section 5.10](https://tools.ietf.org/html/rfc7252#section-5.10). Optimized cache key calculations based on other mechanisms (e.g., using the ETAG) are not included in this PR (and firstly require a proper cache<->proxy integration).

This proposed implementation uses a simple LRU strategy to replace cache entries as soon as the available cache space is used up. The replacement strategy is passed as a function pointer, so that we can easily add more sophisticated strategies in the future, if needed.

~~The demonstrated integration into `gcoap` is preliminary .. depending on the direction we take in #13790, the included changes to `gcoap` may change.~~
The cache implementation now integrates into the forward proxy in #13790.

### Testing procedure

1. unittests are available that show the interaction with the cache
2. To test manually on native: 
   1. build two `gcoap` examples, where one of them uses `USEMODULE=nanocoap_cache`
   2. on the native instance that does not include the `nanocoap_cache` module, type
   ```
    coap get fe80::bc63:a1ff:fe23:df98 5683 /cli/stats
   ```
   the result should be `0`. The **caching** node (our fictitious proxy) caches the generated response.
   3. on the same node, type
   ```
   coap put fe80::bc63:a1ff:fe23:df98 5683 /cli/stats 1
   coap get fe80::bc63:a1ff:fe23:df98 5683 /cli/stats
   ```
   The result should still be `0`, because the caching node does not invoke the request handler function, but performs a cache lookup.
   4. After 60 seconds, repeat the above `get` request, and the response should now contain `1`.

The default `max-age` (see [RFC7252, Section 5.10.5](https://tools.ietf.org/html/rfc7252#section-5.10.5)) is 60 seconds, if no explicit `max-age options` are used.

### Issues/PRs references
~~depends on #13893~~
~~depends #13790~~
